client/src/controllers/KeyboardController.ts

```diff
@@ -0,0 +1,32 @@
+import { Controller } from '@hotwired/stimulus';
+import Mousetrap from 'mousetrap';
+
+/**
+ * Adds the ability to trigger a button click event using a keyboard shortcut
+ * declared on the controlled element.
+ *
+ * @example
+ * ```html
+ * <button type="button" data-controller="w-kbd" data-w-kbd-key="[">Trigger me with the <kbd>[</kbd> key.</button>
+ * ```
+ */
+export class KeyboardController extends Controller<HTMLButtonElement> {
+  static values = { key: { default: '', type: String } };
+
+  handleKey(event: Event) {
+    if (event.preventDefault) event.preventDefault();
+    this.element.click();
+  }
+
+  initialize() {
+    this.handleKey = this.handleKey.bind(this);
```

Comment:
> In initalize() of the KeyboardController class, why are we binding this to this.handleKey? Shouldn't its this property be set to the object by default?

Great question, this could probably work without the binding so maybe check.

However, normally for handlers for events like this its safer to pass a bound version of the function.

What we want is to ensure that no matter how `handleKey` is called, it will have a reference to `this` that is the 'instance of the Controller'. Without doing this bind, we risk the `handleKey` having a `this` which may be the 'keyboard event'.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this - is a good starting point for more reading.


----

The other approach we could take, is to change the `Mousetrap.bind` call as follows;

```
Mousetrap.bind(key, (..._) => this.handleKey(...));
```

But this may make TypeScript get confused and also 'creates' a new anonymous function each time the `keyValueChanged` gets run and also means debugging will be harder (depending on browser).

Hope that helps explain things a bit.

Reply:
I checked it and it doesn't work without binding. Thanks for the explanation. I kinda understood how it works. 

> In non–strict mode, if a function is called with a `this` value that's not an object, the `this` value is substituted with an object

I think the above line in the doc explains it well